Spigot-Server-Patches/0369-Implement-an-API-for-CanPlaceOn-and-CanDestroy-NBT-v.patch

```diff
@@ -0,0 +1,143 @@
+         }
++        // Paper start - Implement an API for CanPlaceOn and CanDestroy NBT values
++        if(tag.hasKey(CAN_DESTROY.NBT)) {
++            NBTTagList list = tag.getList(CAN_DESTROY.NBT, CraftMagicNumbers.NBT.TAG_STRING);
++            for(int i = 0; i < list.size(); i++) {
++                Material material = Material.matchMaterial(list.getString(i), false);
++                this.canDestroy.add(material);
++            }
++        }
++
++        if(tag.hasKey(CAN_PLACE_ON.NBT)) {
++            NBTTagList list = tag.getList(CAN_PLACE_ON.NBT, CraftMagicNumbers.NBT.TAG_STRING);
++            for(int i = 0; i < list.size(); i++) {
++                Material material = Material.matchMaterial(list.getString(i), false);
++                this.canDestroy.add(material);
++            }
++        }
++        // Paper end
+ 
+         if (tag.get(ATTRIBUTES.NBT) instanceof NBTTagList) {
+             NBTTagList save = null;
+@@ -630,6 +661,25 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable {
+         if (hasDamage()) {
+             itemTag.setInt(DAMAGE.NBT, damage);
+         }
++        // Paper start - Implement an API for CanPlaceOn and CanDestroy NBT values
++        if(!this.canPlaceOn.isEmpty()) {
++            List<String> items = this.canPlaceOn.stream()
++                .map(Material::getKey)
++                .map(org.bukkit.NamespacedKey::toString)
++                .collect(java.util.stream.Collectors.toList());
++
++            itemTag.set(CAN_PLACE_ON.NBT, createStringList(items));
++        }
++
++        if(!this.canDestroy.isEmpty()) {
++            List<String> items = this.canDestroy.stream()
++                .map(Material::getKey)
++                .map(org.bukkit.NamespacedKey::toString)
++                .collect(java.util.stream.Collectors.toList());
++
++            itemTag.set(CAN_DESTROY.NBT, createStringList(items));
++        }
++        // Paper end
+ 
+         for (Map.Entry<String, NBTBase> e : unhandledTags.entrySet()) {
+             itemTag.set(e.getKey(), e.getValue());
+@@ -1103,7 +1153,9 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable {
+                         CraftMetaArmorStand.NO_BASE_PLATE.NBT,
+                         CraftMetaArmorStand.SHOW_ARMS.NBT,
+                         CraftMetaArmorStand.SMALL.NBT,
+-                        CraftMetaArmorStand.MARKER.NBT
++                        CraftMetaArmorStand.MARKER.NBT,
++                        CAN_DESTROY.NBT,
++                        CAN_PLACE_ON.NBT
+                         // Paper end
+                 ));
+             }
+@@ -1150,4 +1202,33 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable {
+         return spigot;
+     }
+     // Spigot end
++    // Paper start - Implement an API for CanPlaceOn and CanDestroy NBT values
++    @Override
++    public Set<Material> getCanDestroy() {
++        return java.util.Collections.unmodifiableSet(canDestroy);
```

Comment:
Lets return a normal hashset here. Still be a copy, just mutable

otherwise using the api will require making yet another copy.